### PR TITLE
Reminder screening emails

### LINKED ISSUE #2793: Event Reminder emails
- [ ] Create automated reminder emails for events 

- is it possible to differentiate screenings and other events ? 
- and is it possible to add the timeframe set by the sales agent for the screening as a variable specified in the email ? to specifiy "you have ((timeframe)) to launch the film"

### PR DESCRIPTION
**Pull request by Mussuranae**
Send 2 emails : first, one day before the event starts, second, one hour before the event starts
Fix #2793

### CODE REVIEW

File: apps/backend-functions/src/internals/invitations/events.ts [hunk 1]
```diff
@@ -272,43 +273,73 @@ export async function onInvitationToAnEventUpdate(
...
 
-  // 2 Fetch attendees (invitations accepted)
-  const eventIds: string[] = eventsCollection.docs.map(doc => doc.data().id);
-  const promises = eventIds.map(id => db.collection('invitations').where('eventId', '==', id).where('status', '==', 'accepted').get());
-  const invitationsSnaps = await Promise.all(promises);
-  const invitations: InvitationDocument[] = [];
+    const eventsHourCollection = await db.collection('events')
+    .where('start', '>=', new Date(Date.now() + oneHour))
+    .where('start', '<', new Date(Date.now() + oneHour + halfHour))
+    .get();
```
Comment (phpgeekfr) [inline apps/backend-functions/src/internals/invitations/events.ts:293]: Maybe yan you create a function  like  
```
fetchEventsStaringIn(from: number, to: number) {
  return await db.collection('events')
      .where('start', '>=', new Date(Date.now() + from))
       .where('start', '<', new Date(Date.now() +to))
       .get();
}
```

File: apps/backend-functions/src/internals/invitations/events.ts [hunk 2]
```diff
@@ -272,43 +273,73 @@ export async function onInvitationToAnEventUpdate(
...
+  const eventHourIds: string[] = eventsHourCollection.docs.map(doc => doc.data().id);
+  const dayPromises = eventDayIds.map(id => db.collection('invitations').where('eventId', '==', id).where('status', '==', 'accepted').get());
+  const hourPromises = eventHourIds.map(id => db.collection('invitations').where('eventId', '==', id).where('status', '==', 'accepted').get());
+  const invitationsDaySnaps = await Promise.all(dayPromises);
+  const invitationsHourSnaps = await Promise.all(hourPromises);
+  const invitationsDay: InvitationDocument[] = [];
+  const invitationsHour: InvitationDocument[] = [];
+
+  invitationsDaySnaps.forEach(snap => snap.docs.forEach(doc => invitationsDay.push(doc.data() as InvitationDocument)));
+  invitationsHourSnaps.forEach(snap => snap.docs.forEach(doc => invitationsHour.push(doc.data() as InvitationDocument)));
```
Comment (phpgeekfr) [inline apps/backend-functions/src/internals/invitations/events.ts:306]: Same, create a function that takes `eventsDayCollection.docs` or `eventsHourCollection.docs`
and returns invitations

File: apps/backend-functions/src/internals/invitations/events.ts [hunk 3]
```diff
@@ -272,43 +273,73 @@ export async function onInvitationToAnEventUpdate(
...
+
+    // There is no existing notification for this user
+    if (existingNotifications.empty) {
+      notifications.push(createNotification({
+        toUserId,
+        docId: invitation.eventId,
+        type: 'oneDayReminder'
+      }));
+    }
+  }
```
Comment (phpgeekfr) [inline apps/backend-functions/src/internals/invitations/events.ts:326]: Same, create a function that takes invitations ( invitationsDay or invitationsHour) and notificationType as input

File: apps/backend-functions/src/notification.ts [hunk 4]
```diff
@@ -65,13 +69,39 @@ export function createNotification(notification: Partial<NotificationDocument> =
...
+
+        const template = reminderEventToUser(movie.title.international, user, orgName(org), eventData, templateIds.eventReminder.oneHour);
+        await sendMailFromTemplate(template, 'festival') // TODO guess appKey
+        .then(_ => { notification.email.isSent = true })
+        .catch(e => { notification.email.error = e.message });
+      }
+
       // Send email
       // const appKey = await getOrgAppKey(user.orgId); //@TODO also use notification.type to guess appKey
       // await sendMailFromTemplate({ to: user.email, templateId: 'TODO#4046', data: {} }, appKey); // @TODO #4046
```
Comment (phpgeekfr) [inline apps/backend-functions/src/notification.ts:107]: Transform into a switch/case with 

```
    await sendMail({ to: user.email, subject: notification.type, text: 'test' })
    .then(_ => notification.email.isSent = true)
    .catch(e => notification.email.error = e.message);
  ``` 
  as try catch default

File: apps/backend-functions/src/notification.ts [hunk 5]
```diff
@@ -65,13 +69,39 @@ export function createNotification(notification: Partial<NotificationDocument> =
...
+        // Send reminder email to invited users a day before event start
+        const event = await getDocument<EventDocument<Screening>>(`events/${notification.docId}`);
+        const org = await getDocument<OrganizationDocument>(`orgs/${event.ownerId}`);
+        const eventData = getEventEmailData(event);
+        const movie = await getDocument<MovieDocument>(`movies/${event.meta.titleId}`);
+
+        const template = reminderEventToUser(movie.title.international, user, orgName(org), eventData, templateIds.eventReminder.oneDay);
+        await sendMailFromTemplate(template, 'festival') // TODO guess appKey
+        .then(_ => { notification.email.isSent = true })
+        .catch(e => { notification.email.error = e.message });
```
Comment (phpgeekfr) [inline apps/backend-functions/src/notification.ts:89]: you can remove `{` and  `}`

````
  .then(_ => notification.email.isSent = true )
     .catch(e => notification.email.error = e.message );
  ```

File: apps/backend-functions/src/notification.ts [hunk 6]
```diff
@@ -65,13 +69,39 @@ export function createNotification(notification: Partial<NotificationDocument> =
...
+        // Send reminder email to invited users a day before event start
+        const event = await getDocument<EventDocument<Screening>>(`events/${notification.docId}`);
+        const org = await getDocument<OrganizationDocument>(`orgs/${event.ownerId}`);
+        const eventData = getEventEmailData(event);
+        const movie = await getDocument<MovieDocument>(`movies/${event.meta.titleId}`);
+
+        const template = reminderEventToUser(movie.title.international, user, orgName(org), eventData, templateIds.eventReminder.oneHour);
+        await sendMailFromTemplate(template, 'festival') // TODO guess appKey
+        .then(_ => { notification.email.isSent = true })
+        .catch(e => { notification.email.error = e.message });
```
Comment (phpgeekfr) [inline apps/backend-functions/src/notification.ts:102]: only difference with above is templateId, create a function to not repeat yourself.
Then in the switchCase you can do
```
      switch (notification.type) {
        case 'eventIsAboutToStart':
          yourNewFunction(templateIds.eventReminder.oneHour)
          break
        case 'oneDayReminder':
         yourNewFunction(templateIds.eventReminder.oneDay)
          break;
      
        default:

          await sendMail({ to: user.email, subject: notification.type, text: 'test' })
          .then(_ => notification.email.isSent = true)
          .catch(e => notification.email.error = e.message);
          break;
      }
```

For `await sendMailFromTemplate(template, 'festival') // TODO guess appKey`
we have user  so you can use `getOrgAppKey`

File: libs/notification/src/lib/+state/notification.firestore.ts [hunk 7]
```diff
@@ -26,6 +26,7 @@ export type NotificationType =
   // Events related notifications
   'requestToAttendEventSent' |
   'eventIsAboutToStart' |
```
Comment (phpgeekfr) [inline libs/notification/src/lib/+state/notification.firestore.ts:28]: please add comment
 // 1h (or 15 min ?) efore event start
// 24h before event start

File: package-lock.json [hunk 8]
```diff
@@ -10847,9 +10847,9 @@
           "dev": true
         },
```
Comment (phpgeekfr) [inline package-lock.json:10848]: you pushed package-lock.json but no changes in package.json, can you remove from commit ?

**Review by phpgeekfr — CHANGES_REQUESTED** [review]

**Review by phpgeekfr — APPROVED** [review]

File: apps/backend-functions/src/internals/invitations/events.ts [hunk 9]
```diff
@@ -272,29 +273,60 @@ export async function onInvitationToAnEventUpdate(
 
 /**
  * Notify users with accepted invitation to events that are starting in 1 day
+ * Send also a reminder email
  */
 export async function createNotificationsForEventsToStart() {
   const db = admin.firestore();
```
[the remark this answers is not in the document]
Comment (Mussuranae) [reply apps/backend-functions/src/internals/invitations/events.ts:279]: Nop, good catch !

**Pull request merged**